[RFC] write .nvimlog to $XDG_DATA_HOME/nvimlog

CosmicToast opened the pull request:
Log file is now saved into $XDG_DATA_HOME/nvimlog.

Unfortunately, we recalculate where that is every time we want to open the log file, but this shouldn't be a big problem (the expensive part is the IO operation of opening the file, the calculation itself is trivial (+ a small realloc)).

Commit:
Fix #3561

Log file is now saved into $XDG_DATA_HOME/nvimlog.
Unfortunately, we recalculate where that is every time we want to open the log file, but this shouldn't be a big problem (the expensive part is the IO operation of opening the file, the calculation itself is trivial (+ a small realloc)).

Commit:
prevent memory leak in open_log_file

accidentally let a memory leak slip in, gone now
also simplified procedure filenames (should pass clint test now)
notes: we still end up re-calculating the dir every time (including expensive memory allocations), potentially read more into stdpaths_get_xdg_var to find a more optimal approach

Commit:
fix clint errors

based on output of job 88280584

fmoralesc commented:
Besides using the new location by default, which Is something I agree with, I think the user should be able to override the default location via ENV var, as is possible now. So: try to use the env var, then fallback to the XDG location.

5paceToast commented:
This'll require some deeper digging (which I'll probably do, just not today).
If anyone wants to give me a PR for that I'll gladly accept it (as well as anything that'll help out with the way we need to recalculate this every time).

tarruda commented:
Shouldn't it be `$XDG_DATA_HOME/nvim/log` ?

jrobeson commented:
thanks for this @5paceToast

5paceToast commented:
@tarruda will patch in a bit (I thought that $XDG_DATA_HOME/nvimlog was reasonable, but your suggestion works too)
@fmoralesc any idea what that env variable should be? NVIM_LOG?

Commit:
update default log location

old: $XDG_DATA_HOME/nvimlog
new: $XDG_DATA_HOME/nvim/log

fmoralesc commented:
@5spaceToast That sound good ;)

justinmk commented on src/nvim/log.c:
```diff
@@ -77,20 +76,15 @@ FILE *open_log_file(void)
-      goto open_log_file_error;
-    }
+    char * dir = stdpaths_get_xdg_var(kXDGDataHome);
+    dir = concat_fnames_realloc(dir, "nvim/log", true);
```
this only needs to be done once, in `log_init()`.

justinmk commented on src/nvim/log.c:
```diff
@@ -64,7 +63,7 @@ bool do_log(int log_level, const char *func_name, int line_num, bool eol,
 ///
-/// @return The FILE* specified by the USR_LOG_FILE path or stderr in case of
+/// @return The FILE* to log into or stderr in case of
 ///         error
```
fits on one line

justinmk commented on src/nvim/log.c:
```diff
@@ -77,20 +76,15 @@ FILE *open_log_file(void)
-    static char expanded_log_file_path[MAXPATHL + 1];
-
-    expand_env((char_u *)USR_LOG_FILE, (char_u *)expanded_log_file_path,
-               MAXPATHL);
```
This was only expanding `"$HOME/.nvimlog"`, it was never configurable via env var. Does anyone actually plan to change the location from XDG_DATA_HOME?

fmoralesc replied:
Bah, I thought it was something more interesting for some reason :p I thought that might be useful when debugging?

Commit:
Initialize Log Path Once

Move log file path initialization into log_init.

5paceToast commented:
After latest patch, neovim can no longer open the logfile.
However, the path is properly generated, and is accessible in the scope of fopen.

I'm guessing something may be trying to LOG before intialization? I'm not seeing how this makes any sense, however...

I'll write up another patch for the env variable in a bit, while we try to figure out why it wouldn't be able to open the file.

Commit:
Add support for NVIM_LOG_FILE env variable

Todo: verify that NVIM_LOG_FILE is valid
Note: NVIM_LOG_FILE-specified files can be opened, but the XDG file
cannot, (but can if manually specified).
Looks like the XDG file path isn't being generated properly.

5paceToast commented:
It appears that the XDG file path is not being generated properly (despite looking right when printf'd). I'm guessing a missing `'\0'`?

If specified manually using NVIM_LOG_FILE, putting it into $XDG_DATA_HOME/nvim/log works (specified directly as ~/.local/share/nvim/log, which is what it expands to).

Looking into how I'm generating it.

Commit:
Move XDG logfile detection to stdpaths_user_data_subpath

Commit:
Fix clint style conformity

- spaces after //-style comments
- TODO(username)

Pyrohh commented on src/nvim/log.c:
```diff
@@ -25,6 +26,13 @@ static uv_mutex_t mutex;
 {
   uv_mutex_init(&mutex);
+  //determine where to save the log file
+  log_file_dir = getenv("NVIM_LOG_FILE");
```
You should use `os_getenv()` (see `src/nvim/os/env.c`):
```
/// Like getenv(), but returns NULL if the variable is empty.
const char *os_getenv(const char *name)
  FUNC_ATTR_NONNULL_ALL
{
  const char *e = getenv(name);
  return e == NULL || *e == NUL ? NULL : e;
}
```

5paceToast commented:
Delegated XDG logfile location selection to stdpaths_user_data_subpath (as with ShaDa).
Still cannot open file unless the env variable is present... does anyone have an idea on what may have introduced this regression?

Also, while on my development machine, all tests are passed (with one msgpack functional test error, but not failure), it seems that on, as a minimum, osx 64 it fails some test success conditions.

5paceToast commented:
Checking with errno, the error is 2 (ENOENT, "A component of a specified pathname did not exist, or the pathname was an empty string.").

The string is definitely readable then, but the pathname also does exist...
Waiting to push os_getpath until I think about this some more.

5paceToast commented:
GCONV (of any type) doesn't seem to be liking this (no response after some tests).

5paceToast commented:
I think I found the issue : `'~'` does not expand when used in fopen.
Even if I specify ~/.local (etc) in the env variable, the shell is what expands this.

Looking into how ShaDa opens its file, and thinking of potential solutions.

5paceToast commented:
Fix seems to be simple: `expand_env`.
Going to go through that and run tests.

Commit:
Use os_getenv and expand_env

expand_env seems to always fail.
also, now using static array rather than pointer as in pre-patch

5paceToast commented:
`expand_env` appears to always fail for some reason. RFC?

fmoralesc commented:
I can confirm the issue, I haven't had luck with it either.

5paceToast commented:
Also wondering what the deal is with GCOV and tests hanging. Maybe because the log file can't be opened?

Also, what reason do we have for expand_env being manually written? I'm fairly sure `wordexp.h` is available on POSIX platforms (at least, musl has it, which is a strong suggestion towards that), and on Windows platforms there's `ExpandEnvironmentStrings` and `ExpandEnvironmentStringsForUser`.

5paceToast commented:
Confirming that the issue here is with how `expand_env` is implemented.

I wrote a replacement for `expand_env_esc` (`wordexp.h` for UNIX platforms, a temporary stub for everything else) and `log_init` now successfully detects the proper place to save the log, and opens it without failure.

I'll start trying to port it over properly, would anyone consider helping out with the non-*nix implementations? (I know that `ExpandEnvironmentStrings` and `ExpandEnvironmentStringsForUser` must be used, but I'm less "at home" with MS libs)

justinmk commented:
Closing per https://github.com/neovim/neovim/issues/3561#issuecomment-192809706

@5paceToast Your work on `expand_env_esc` seems like it might be worth pursuing in another PR.
